src/EasyNetQ.HostedService/RabbitMqProducer.cs

```diff
@@ -249,55 +283,65 @@ private protected virtual void OnCancellation()
                             message.TaskCompletionSource.SetResult(PublishResult.Published);
 
                             _messages.TryDequeue(out _);
+
+                            success = true;
                         }
                     }
-                    catch (OperationCanceledException)
+                    catch (OperationCanceledException e)
                     {
-                        Logger?.LogDebug($"Stopping producer loop with configuration {RabbitMqConfig.Id}.");
+                        if (e.CancellationToken == cancellationToken)
```

Comment:
If the cancellation token is different (which means we got a NACK or did not get ACK in time) we want to make sure that we `FailAndDiscardMessage` or somehow flag it for retry which means resetting the `_messageSemaphore` so it will go through it in the next loop iteration